contextualized/dags/torch_notmad/torch_notmad.py

```diff
@@ -0,0 +1,286 @@
+    
+    def forward(self,batch_weights):
+        batch_size = batch_weights.shape[0]
+        archetypes = self._archetypes()
+        
+        batch_archetypes = torch.tensor(np.zeros((batch_size, self.d, self.d)))
+        for i,batch_w in enumerate(batch_weights):
+            batch_archetypes[i] = torch.tensordot(batch_w.float(), archetypes.float(), dims=1)
+        
+        return batch_archetypes
+                
+    
+    def _archetypes(self):
+        return torch.multiply(self.archs, self.mask)
+    
+
+class NOTMAD_model(pl.LightningModule):
+    """
+    NOTMAD model
+    """
+    def __init__(self, 
+                 datamodule,
+                 n_archetypes = 4,
+                 sample_specific_loss_params = {'l1': 0., 'alpha': 1e-1, 'rho': 1e-2},
+                 archetype_loss_params = {'l1': 0., 'alpha': 1e-1, 'rho': 1e-2},
+                 opt_lr = 1e-3,
+                 opt_step= 50,
+                 n_encoder_hidden_layers=2,
+                 encoder_width=32,
+                 init_mat=None,
+                 auto_opt=True,
+                 encoder_type="NGAM"
+            ):
+        """ Initialize NOTMAD.
+        
+        Args:
+            datamodule (pl.LightningDataModule): Lightning datamodule to use for training
+        
+        Kwargs:
+            Explainer Kwargs
+            ----------------
+            init_mat (np.array): 3D Custom initial weights for each archetype. Defaults to None.
+            n_archetypes (int:4): Number of archetypes in explainer
+            
+            Encoder Kwargs
+            ----------------
+            n_encoder_hidden_layers(int:2): Number encoder hidden layers
+            encoder_width(int:32): Width of encoder hidden layers
+            encoder_type (str: NGAM): Encoder module to use
+            
+            Optimization Kwargs
+            -------------------
+            auto_opt(bool: True): Use torch's backprop vs manual (customizable backprop in self.training_step)
+            opt_lr(float): Optimizer learning rate
+            opt_step(int): Optimizer step size
+            
+            Loss Kwargs
+            -----------
+            sample_specific_loss_params (dict of str: int): Dict of params used by NOTEARS loss (l1, alpha, rho)
+            archetype_specific_loss_params (dict of str: int): Dict of params used by Archetype loss (l1, alpha, rho)
+
+        """
+        super(NOTMAD_model, self).__init__()
+
+        #dataset params
+        self.datamodule = datamodule
+        self.n_archetypes = n_archetypes
+        self.context_shape = self.datamodule.C_train.shape
+        self.feature_shape = self.datamodule.X_train.shape
+
+        #dag/loss params
+        self.project_distance = 0.1
+        self.archetype_loss_params = archetype_loss_params
+        self.alpha, self.rho, self.use_dynamic_alpha_rho = self._parse_alpha_rho(sample_specific_loss_params)
+
+        #torch params
+        self.opt_lr = opt_lr
```

Comment:
Needs to be self.learning_rate to enable Torch Lightning's auto_lr_find feature.